packages/dart/sshnoports/lib/src/noports_cli/issue_keys/np_issue_keys.dart

```diff
@@ -0,0 +1,167 @@
+  /// into a format compatible with 'noports activate' command.
+  ///
+  /// Returns: activation string in format `<atsign>:enroll:otp:<otp>[:name:<deviceName>[:keyfile:<path>]]`
+  String _generateEnrollCommand(NoportsParams params) {
+    StringBuffer cbuf = StringBuffer();
+    cbuf.append(NPIssueKeys.baseEnrollCommand
+        .replaceFirst('<atsign>', params.atsign)
+        .replaceFirst('<otp>', params.otp!));
+
+    // write optional args if available
+    cbuf.append('[:name:${params.deviceName}');
+    if (!(params.atKeysFilePath == null || params.atKeysFilePath == '')) {
+      cbuf.append(':keyfile:${params.atKeysFilePath}');
+    }
+    cbuf.append(']');
+
+    return cbuf.message!;
+  }
+
+  /// Approves the first pending enrollment request with inferred noports parameters
+  ///
+  /// Creates an ApprovedRequestDecision and submits it to the enrollment service.
+  ///
+  /// Throws: [AtEnrollmentException] if enrollment approval fails
+  /// Returns: [AtEnrollmentResponse] containing the enrollment status
+  Future<AtEnrollmentResponse> _approveFirstPendingEnrollment(
+      NoportsParams params) async {
+    Enrollment enrollment = await _awaitAndFetchEnrollmentRequest(params);
+    writeInfoMessage(
+        'Approving enrollment with id: ${enrollment.enrollmentId}');
+
+    ApprovedRequestDecisionBuilder decisionBuilder =
+        ApprovedRequestDecisionBuilder(
+            enrollmentId: enrollment.enrollmentId!,
+            encryptedAPKAMSymmetricKey: enrollment.encryptedAPKAMSymmetricKey!);
+    EnrollmentRequestDecision decision =
+        EnrollmentRequestDecision.approved(decisionBuilder);
+
+    AtEnrollmentResponse response = await _enrollmentService!.approve(decision);
+
+    return response;
+  }
+
+  /// Waits for and fetches the first pending enrollment request matching [params]
+  ///
+  /// Polls the enrollment service at regular intervals until an enrollment request
+  /// is found with matching deviceName, appName, and namespace.
+  ///
+  /// Throws: [AtEnrollmentException] if enrollmentId is missing from the response
+  /// Returns: The first pending [Enrollment] that matches the criteria
+  Future<Enrollment> _awaitAndFetchEnrollmentRequest(
+      NoportsParams params) async {
+    writeInfoMessage('Waiting for enrollment request - '
+        'will retry every ${enrollmentCheckInterval / 1000}s');
+    EnrollmentListRequestParam enrollListParams = EnrollmentListRequestParam()
+      ..deviceName = params.deviceName
+      ..appName = defaultAppName
+      ..namespace = defaultEnrollmentNamespaces.toString()
+      ..enrollmentListFilter = [EnrollmentStatus.pending];
+
+    writeInfoMessage('Listening...');
+    while (true) {
+      List<Enrollment> enrollRequests = await _enrollmentService!
+          .fetchEnrollmentRequests(enrollmentListParams: enrollListParams);
+
+      if (enrollRequests.isEmpty) {
+        // keeps the loop active until enrollment fetched or terminated
```

Comment:
Currently keeps retrying forever until successful or terminated. Should we consider having a retry limit?